test/integration/auth/dynamic_client_test.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authorization/authorizerfactory"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	utilfeaturetesting "k8s.io/apiserver/pkg/util/feature/testing"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/features"
+	kubeoptions "k8s.io/kubernetes/pkg/kubeapiserver/options"
+	"k8s.io/kubernetes/pkg/master"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestDynamicClientBuilder(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.TokenRequest, true)()
+
+	tmpfile, err := ioutil.TempFile("/tmp", "key")
+	if err != nil {
+		t.Fatalf("create temp file failed: %v", err)
+	}
+	defer os.RemoveAll(tmpfile.Name())
+
+	if err = ioutil.WriteFile(tmpfile.Name(), []byte(ecdsaPrivateKey), 0666); err != nil {
+		t.Fatalf("write file %s failed: %v", tmpfile.Name(), err)
+	}
+
+	const iss = "https://foo.bar.example.com"
+	aud := authenticator.Audiences{"api"}
+
+	maxExpirationSeconds := int64(60 * 60)
+	maxExpirationDuration, err := time.ParseDuration(fmt.Sprintf("%ds", maxExpirationSeconds))
+	if err != nil {
+		t.Fatalf("parse duration failed: %v", err)
+	}
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	baseClient, baseConfig := framework.StartTestServer(t, stopCh, framework.TestServerSetup{
+		ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
+			opts.ServiceAccountSigningKeyFile = tmpfile.Name()
+			opts.ServiceAccountTokenMaxExpiration = maxExpirationDuration
+			if opts.Authentication == nil {
+				opts.Authentication = &kubeoptions.BuiltInAuthenticationOptions{}
+			}
+
+			opts.Authentication.APIAudiences = aud
+			if opts.Authentication.ServiceAccounts == nil {
+				opts.Authentication.ServiceAccounts = &kubeoptions.ServiceAccountAuthenticationOptions{}
+			}
+			opts.Authentication.ServiceAccounts.Issuer = iss
+			opts.Authentication.ServiceAccounts.KeyFiles = []string{tmpfile.Name()}
+		},
+		ModifyServerConfig: func(config *master.Config) {
+			config.GenericConfig.Authorization.Authorizer = authorizerfactory.NewAlwaysAllowAuthorizer()
+		},
+	})
+
+	// We want to test if the token rotation works fine here.
+	// To minimize the time this test would consume, we use the minimial token expiration.
+	// The minimial token expiration is defined in:
+	// pkg/apis/authentication/validation/validation.go
+	exp := int64(600)
+	leeway := 80
+	ns := "default"
+	clientBuilder := controller.NewTestDynamicClientBuilder(
+		restclient.AnonymousClientConfig(baseConfig),
+		baseClient.CoreV1(),
+		baseClient.AuthenticationV1(),
+		ns, exp, leeway)
+
+	saName := "dt"
+	dymClient, err := clientBuilder.Client(saName)
+
+	if err != nil {
+		t.Fatalf("build client via dynamic client builder failed: %v", err)
+	}
+
```

Comment:
just waiting isn't enough... we need to make sure the token actually gets renewed.

you can do this like this:

```
if inErr := testClientBuilder(dymClient, ns, saName); inErr != nil {
	t.Fatalf("dynamic get resources failed: %v", inErr)
}
... delete the service account, invalidating the token ...
time.Sleep(time.Second * 10)
if inErr := testClientBuilder(dymClient, ns, saName); inErr != nil {
	t.Fatalf("dynamic get resources failed: %v", inErr)
}
```

Reply:
that also avoids a goroutine in the test